ioSPI/fourier.py

```diff
@@ -0,0 +1,212 @@
+import numpy as np
+import pyfftw.interfaces.numpy_fft
+from numba import jit
+
+
+@jit
+def neg_pos_2d(arr2d):
+    """Apply checkerboard pattern.
+
+    Each pixel switches from positive to negative in
+    checker board pattern.
+
+    Parameters
+    ----------
+    arr2d : numpy.ndarray, shape (N,N)
+        Input array.
+
+    Returns
+    -------
+    arr2d : numpy.ndarray, shape (N,N)
+        Output array.
+    """
+    assert arr2d.ndim == 3  # extra axis
+    for n_particle in range(arr2d.shape[0]):
+        for r in range(arr2d.shape[1]):
+            for c in range(arr2d.shape[2]):
+                if (r + c) % 2:
+                    arr2d[n_particle, r, c] *= -1
+    return arr2d
+
+
+@jit
+def make_neg_pos_3d(arr3d):
+    """Apply checkerboard pattern.
+
+    Each pixel switches from positive to negative in
+    checker board pattern.
+
+    Parameters
+    ----------
+    arr3d : numpy.ndarray, shape (N,N,N)
+        Input array.
+
+    Returns
+    -------
+    neg_pos_3d : numpy.ndarray, shape (N,N,N)
+        Output array.
+    """
+    R1, R2, R3 = arr3d.shape
+    neg_pos_3d = np.ones(arr3d.shape)
+    for r1 in range(R1):
+        for r2 in range(R2):
+            for r3 in range(R3):
+                if (r1 + r2 + r3) % 2:
+                    neg_pos_3d[r1, r2, r3] *= -1.0
+    return neg_pos_3d
+
+
+def fft3d(
+    arr3d,
+    mode,
+    neg_pos_3d=None,
+    numpy_fft=pyfftw.interfaces.numpy_fft,
+    only_real=False
+):
+    """3D FFT
+
+    Fast fourier transform
+
+    Parameters
+    ----------
+    arr3d : numpy.ndarray, shape (N,N,N)
+        Input array.
+    mode : str
+        Forward or reverse
+    neg_pos_3d : numpy.ndarray, shape (N,N,N)
+        Optinal precomputed array. If None (default) passed,
+        computes on the fly.
+    numpy_fft : func
+        Function backend for performing the fft.
+    only_real : bool
+        If True, return only real part of transform.
+
+    Returns
+    -------
+    arr3d_f : numpy.ndarray, shape (N,N,N)
+        Output array.
+    """
+    if neg_pos_3d is None:
+        neg_pos_3d = make_neg_pos_3d(arr3d)
+    if arr3d.shape[0] % 4 != 0:
```

Comment:
Maybe add a comment on what this is doing?

Reply:
I want to put this ref in the docstring, but its large. Should I wrap it with a \ ?
https://dsp.stackexchange.com/questions/9039/centering-zero-frequency-for-discrete-fourier-transform

Reply:
Thanks Geoff, I learned something! I still need to wrap my head around this but this is really cool

Reply:
maybe use a tinyurl?

Reply:
it's actually not causing problems. Maybe when there are no spaces, the 79 characters per line is not enforced.